### PR TITLE
feat(firestore, types): support Generic Types

### PR DESCRIPTION
**Pull request by franciscosolla**
Updates firestore types to generic types, as is in the firestore web SDK.
Allows to define specific document data types ensuring type safe operations with collections and documents, plus the convenience of automatic auto-complete from type defined document data.

### Description

I use type defined document data with firebase in my projects. Because the react-native-firebase implementation of firestore is not type generic I find myself replicating the type file to make then generic in my projects. As I have seen other people with the same problem, I decided to commit the types I've been using.

### Checklist

- I read the [Contributor Guide](/CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [ ] Yes
- My change supports the following platforms;
  - [X] `Android`
  - [X] `iOS`
- My change includes tests;
  - [ ] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [X] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [X] No

### Test Plan

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

### CODE REVIEW

**Comment by mikehardy** [comment]
Cool! Has a CI fail though, which I think would be resolved by simply opening the root of the monorepo in Visual Studio Code, re-opening the file, and asking it to reformat. Or perhaps from command-line asking prettier to auto-fix? I have it run as a save-hook on Code now so I'm not sure if it's configured correctly for command line

```

/home/runner/work/react-native-firebase/react-native-firebase/packages/firestore/lib/index.d.ts
##[error]    93:5   error  Replace `··[key:·string]:·DocumentFieldType` with `[key:·string]:·DocumentFieldType;`                                                                                                                      prettier/prettier
##[error]   554:80  error  Insert `⏎···`                                                                                                                                                                                              prettier/prettier
##[error]  1319:13  error  Replace `callback:·(result:·QueryDocumentSnapshot<T>,·index:·number)·=>·void,·thisArg?:·any` with `⏎······callback:·(result:·QueryDocumentSnapshot<T>,·index:·number)·=>·void,⏎······thisArg?:·any,⏎····`  prettier/prettier
##[error]  1552:48  error  Replace `documentRef:·DocumentReference<T>` with `⏎······documentRef:·DocumentReference<T>,⏎····`                                                                                                          prettier/prettier
##[error]  1605:51  error  Replace `documentRef:·DocumentReference<T>,·data:·Partial<T>` with `⏎······documentRef:·DocumentReference<T>,⏎······data:·Partial<T>,⏎····`                                                                prettier/prettier
##[error]  1727:51  error  Replace `documentRef:·DocumentReference<T>,·data:·T` with `⏎······documentRef:·DocumentReference<T>,⏎······data:·T,⏎····`                                                                                  prettier/prettier
##[error]  1853:55  error  Replace `collectionPath:·string` with `⏎······collectionPath:·string,⏎····`                                                                                                                                prettier/prettier
```

**Comment by mikehardy** [comment]
Hi @franciscosolla ! Sorry for the delay in reviewing this. I like the change a lot, I'm leaning towards merging it but I'd like to ask - are you certain this is not a breaking change? Or will this break typings for any of the users?

I have no problem with breaking changes, but I do have a problem with accidentally breaking people without it being a semantic-version major release is all, and I want to mark it correctly.

I'll try to test this locally in my project as well to cross-check

Thanks!

**Comment by mikehardy** [comment]
We have added a feature to the repo where a patch-package file is generated for any changes so that others may easily test, the patches that implement this change are here: https://github.com/invertase/react-native-firebase/suites/1145045506/artifacts/16455376

I will test them myself to see if the typings still work in my work project, without regression

File: packages/firestore/lib/index.d.ts [hunk 1]
```diff
@@ -86,11 +86,18 @@ export namespace FirebaseFirestoreTypes {
     toUint8Array(): Uint8Array;
   }
 
+  /**
+   * A `DocumentData` object represents the data in a document.
+   */
+  export interface DocumentData {
+    [key: string]: DocumentFieldType;
```
Comment (mikehardy) [inline packages/firestore/lib/index.d.ts:93]: I integrated the changes into my work project for testing and found them to be breaking
They diverge from the web SDK which is the basis of comparison for the APIs and typing here: https://github.com/firebase/firebase-js-sdk/blob/f506e18facf8f47eff61cf83a98af4b1f2e94319/packages/firestore/exp-types/index.d.ts#L23

```suggestion
    [key: string]: any;
```

When I use this suggestion, it works though, non-breaking, and the types seem to match the upstream 

I will apply this suggestion then this is good to go

**Review by mikehardy — APPROVED** [review]
With the change applied to harmonize the DocumentData type with firebase-js-sdk this seems to work fine in my work project

**Comment by franciscosolla** [comment]
Wow, I had forgotten about this haha, sorry. Thank you for fixing it. I'm happy that I could contribute to your project. I use it in every project I develop haha. Awesome work! @mikehardy

**Pull request merged**